### PR TITLE
GAWB-1473: Allow owners to grant share permissions to others

### PR DESCRIPTION
**Pull request by MatthewBemis**
firecloud-ui: https://github.com/broadinstitute/firecloud-ui/pull/643
firecloud-orch: broadinstitute/firecloud-orchestration#352

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/resources/swagger/rawls.yaml [hunk 1]
```diff
@@ -4277,26 +4279,63 @@ definitions:
...
         type: string
         description: email address of the user or group whose permissions will be changed
       accessLevel:
         type: string
         description: The access level to grant to this user or group (OWNER, READER, WRITER, NO ACCESS)
+      canShare:
+        type: boolean
+        description: Set to true if you want this user to be able to share the workspace with other users
+
+  WorkspaceACLUpdateResponseList:
```
Comment (MatthewBemis) [inline core/src/main/resources/swagger/rawls.yaml:4294]: this particular addition was forgotten in the firecloud invites PR. i did make one small change to this response- i differentiate invitesSent from invitesUpdated so orchestration doesn't potentially send out too many invites. small issue so i fixed it here. orch has been updated accordingly

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 2]
```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
...
+
+      userPermissionQuery.result.map(_.map { case (access, email, canShare) =>
+        val accessLevel = WorkspaceAccessLevels.withName(access)
+        // Sharing is implied for owners and project owners, so the permission may not be in the DB. Rectify that here
+        (email, accessLevel, ((accessLevel >= WorkspaceAccessLevels.Owner) || canShare))
+      }).flatMap { userResults =>
+        groupPermissionQuery.result.map(_.map { case (access, email, canShare) =>
+          val accessLevel = WorkspaceAccessLevels.withName(access)
+          // Sharing is implied for owners and project owners, so the permission may not be in the DB. Rectify that here
+          (email, accessLevel, ((accessLevel >= WorkspaceAccessLevels.Owner) || canShare))
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: to expand on the comment, i deemed it too much overhead to maintain share records in the DB for access levels that implicitly had the ability to share. this would be particularly annoying for project owners, who are altered outside of the context of a workspace (and doing so may impact many workspaces)

this also means we don't need to write any sort of migration

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -195,14 +195,16 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
-              RequestComplete(StatusCodes.OK, WorkspaceListResponse(accessLevel, workspaceContext.workspace, stats, owners))
+          getUserSharePermissions(workspaceContext, accessLevel, dataAccess) flatMap { canShare =>
+            if (accessLevel < WorkspaceAccessLevels.Read)
+              DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.NotFound, noSuchWorkspaceMessage(workspaceName))))
+            else {
+              for {
+                stats <- getWorkspaceSubmissionStats(workspaceContext, dataAccess)
+                owners <- getWorkspaceOwners(workspaceContext.workspace, dataAccess)
+              } yield {
+                RequestComplete(StatusCodes.OK, WorkspaceResponse(accessLevel, canShare, workspaceContext.workspace, stats, owners))
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:206]: I changed this response type because it was also used in the listWorkspaces API call. I needed to return the canShare value but calculating it for the entire list of workspaces was useless and would add a lot of extra queries. So I broke them apart

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 4]
```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
     def deleteWorkspaceInvites(workspaceId: UUID) = {
       findWorkspaceInvitesQuery(workspaceId).delete
     }
-    
-    def listEmailsAndAccessLevel(workspaceContext: SlickWorkspaceContext): ReadAction[Seq[(String, WorkspaceAccessLevel)]] = {
+
+    def updateUserSharePermissions(workspaceId: UUID, users: Map[RawlsUserRef, Option[Boolean]]) = {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: these update functions are ugly. I would just have an insert and a delete since there isn't actually an update.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 5]
```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
...
+        workspaceGroupShareQuery.filter(_.workspaceId === workspaceId).delete
+    }
+
+    def getUserSharePermissions(subjectId: RawlsUserSubjectId, workspaceContext: SlickWorkspaceContext): ReadAction[Boolean] = {
+      workspaceUserShareQuery.filter(rec => (rec.userSubjectId === subjectId.value && rec.workspaceId === workspaceContext.workspaceId)).countDistinct.result.map(rows => rows > 0) flatMap { canShare =>
+        if(canShare) DBIO.successful(canShare)
+        else rawlsGroupQuery.listGroupsForUser(RawlsUserRef(subjectId)).flatMap { userGroups =>
+          val groupNames = userGroups.map(_.groupName.value)
+          workspaceGroupShareQuery.filter(rec => (rec.groupName) inSet(groupNames)).countDistinct.result.map(rows => rows > 0) flatMap { canShareViaGroup =>
+            DBIO.successful(canShareViaGroup)
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: do you need the last flatmap?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 6]
```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
...
-        (email, WorkspaceAccessLevels.withName(access))
-      })
+      } yield (access, subGroup)).map { case (access, subGroup) => (access.accessLevel, subGroup.groupEmail, subGroup.groupName) }
+
+      /*  The left join here is important. Since we are only going to store share-permission records for users and groups that have been
+          explicitly granted that ability, we want to be able to return all users and groups even if they don't have share permissions.
+          After the join, a null column equates to false, and a non-null column equates to true. This conversion is done with permission.isDefined
+       */
+      val userPermissionQuery = accessAndUserEmail.joinLeft(workspaceUserShareQuery).on(_._3 === _.userSubjectId).map { case (userInfo, permission) => (userInfo._1, userInfo._2, permission.isDefined) }
+      val groupPermissionQuery = accessAndSubGroupEmail.joinLeft(workspaceGroupShareQuery).on(_._3 === _.groupName).map { case (groupInfo, permission) => (groupInfo._1, groupInfo._2, permission.isDefined) }
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: use the term Share instead of Permission

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 2 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: why do we have this logic here instead of in requireSharePermission? This makes it feel like there is an actual row in the db which is maybe changeable. I think I would prefer requireSharePermission have the logic that allows access if the user is >= Owner or canShare

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 7]
```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
...
+        else rawlsGroupQuery.listGroupsForUser(RawlsUserRef(subjectId)).flatMap { userGroups =>
+          val groupNames = userGroups.map(_.groupName.value)
+          workspaceGroupShareQuery.filter(rec => (rec.groupName) inSet(groupNames)).countDistinct.result.map(rows => rows > 0) flatMap { canShareViaGroup =>
+            DBIO.successful(canShareViaGroup)
+          }
+        }
+      }
+    }
+
+    def listEmailsAndAccessLevel(workspaceContext: SlickWorkspaceContext): ReadAction[(Seq[(String, WorkspaceAccessLevel, Boolean)], Seq[(String, WorkspaceAccessLevel, Boolean)])] = {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: why does this return a tuple now?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 8]
```diff
@@ -435,15 +438,17 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
           dataAccess.workspaceQuery.listEmailsAndAccessLevel(workspaceContext).flatMap { emailsAndAccess =>
             dataAccess.workspaceQuery.getInvites(workspaceContext.workspaceId).map { invites =>
               // toMap below will drop duplicate keys, keeping the last entry only
               // sort by access level to make sure higher access levels remain in the resulting map
-              val granted = emailsAndAccess.sortBy { case (_, accessLevel) => accessLevel }.map { case (email, accessLevel) => email -> AccessEntry(accessLevel, false)}
-              val pending = invites.sortBy { case (_, accessLevel) => accessLevel }.map { case (email, accessLevel) => email -> AccessEntry(accessLevel, true)}
+              val grantedUsers = emailsAndAccess._1.sortBy { case (_, accessLevel, _) => accessLevel }.map { case (email, accessLevel, canShare) => email -> AccessEntry(accessLevel, false, canShare) }
+              val grantedGroups = emailsAndAccess._2.sortBy { case (_, accessLevel, _) => accessLevel }.map { case (email, accessLevel, canShare) => email -> AccessEntry(accessLevel, false, canShare) }
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: in general I am seeing too much tuple access by position (i.e. x._2). This makes the code less readable. Can you review this PR for these cases and change them to case statements where possible?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 9]
```diff
@@ -481,7 +554,7 @@ trait WorkspaceComponent {
       }
     }
 
-    def findWorkspaceUsersAndAccessLevel(workspaceId: UUID): ReadAction[Set[(Either[RawlsUserRef, RawlsGroupRef], WorkspaceAccessLevel)]] = {
+    def findWorkspaceUsersAndAccessLevel(workspaceId: UUID): ReadAction[Set[(Either[RawlsUserRef, RawlsGroupRef], (WorkspaceAccessLevel, Option[Boolean]))]] = {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: why is the Boolean an Option?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 10]
```diff
@@ -544,20 +562,28 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
+      val refsToUpdate = aclUpdates.map { aclUpdate => (refsToUpdateByEmail.get(aclUpdate.email), aclUpdate.accessLevel, aclUpdate.canShare) }.collect {
+        case (Some(ref), WorkspaceAccessLevels.NoAccess, _) => ref -> (WorkspaceAccessLevels.NoAccess, Option(false))
+        case (Some(ref), accessLevel, canShare) => ref -> (accessLevel, canShare)
       }.toSet
 
+      val refsToUpdateAndSharePermission = refsToUpdate.map(ref => ref._1 -> ref._2._2)
+      val refsToUpdateAndAccessLevel = refsToUpdate.map(ref => ref._1 -> ref._2._1)
+
+      val existingRefsAndSharePermission = existingRefsAndLevels.map(ref => ref._1 -> ref._2._2)
+      val existingRefsAndAccessLevel = existingRefsAndLevels.map(ref => ref._1 -> ref._2._1)
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: :nauseated_face:

**Review by dvoet — CHANGES_REQUESTED** [review]
Maybe adding a new case class that contains both WorkspaceAccessLevel and canShare can help clean things up.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 9 above)
Comment (MatthewBemis) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: So it would be easier to diff this with the user-supplied aclUpdates in determineCompleteNewAcls, but I suppose that would be better handled in determineCompleteNewAcls and not in the database code

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 2 above)
Comment (MatthewBemis) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: this is used in getACL, which now returns acls in the form:

```
[
   {
      "email": "foo@bar.baz",
      "canShare": true,
      "accessLevel": "OWNER"
   }
]
```

it does feel dishonest to be pretending that there are rows in the DB but it felt cleaner code-wise to deal with that here than in getACL

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 11]
```diff
@@ -242,25 +267,77 @@ trait WorkspaceComponent {
...
+        (email, accessLevel, ((accessLevel >= WorkspaceAccessLevels.Owner) || canShare))
+      }).flatMap { userResults =>
+        groupShareQuery.result.map(_.map { case (access, email, canShare) =>
+          val accessLevel = WorkspaceAccessLevels.withName(access)
+          // Sharing is implied for owners and project owners, so the permission may not be in the DB. Rectify that here
+          (email, accessLevel, ((accessLevel >= WorkspaceAccessLevels.Owner) || canShare))
+        }).map { groupResults =>
+          userResults ++ groupResults
+        }
+      }
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:336]: rewrite this last block as a for comprehension

```
      for {
        userResults <- userShareQuery.result
        groupResults <- groupShareQuery.result
      } yield {
        (userResults ++ groupResults).map { case (access, email, hasSharePermission) =>
          (email, WorkspaceAccessLevels.withName(access), hasSharePermission)
        }
      }
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 12]
```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
...
+      } yield (access, subGroup)).map { case (access, subGroup) => (access.accessLevel, subGroup.groupEmail, subGroup.groupName) }
 
-      (accessAndUserEmail union accessAndSubGroupEmail).result.map(_.map { case (access, email) => 
-        (email, WorkspaceAccessLevels.withName(access))
-      })
+      /*  The left join here is important. Since we are only going to store share-permission records for users and groups that have been
+          explicitly granted that ability, we want to be able to return all users and groups even if they don't have share permissions.
+          After the join, a null column equates to false, and a non-null column equates to true. This conversion is done with permission.isDefined
+       */
+      val userShareQuery = accessAndUserEmail.joinLeft(workspaceUserShareQuery).on(_._3 === _.userSubjectId).map { case ((accessLevel, userEmail, _), shareRecord) => (accessLevel, userEmail, shareRecord.isDefined) }
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: The SQL query that slick actually generates:

```
SELECT 
    x2.`access_level`, x3.`EMAIL`, x4.`workspace_id` IS NOT NULL
FROM
    `WORKSPACE_ACCESS` x2
        INNER JOIN
    (`GROUP` x5
    INNER JOIN (`GROUP_USERS` x6
    INNER JOIN `USER` x3 ON x3.`SUBJECT_ID` = x6.`USER_SUBJECT_ID`) ON x5.`NAME` = x6.`GROUP_NAME`) ON ((x2.`workspace_id` = 0xE5D5CA221F0146DF9F92551963A9A1)
        AND (x2.`is_realm_acl` = FALSE))
        AND (x2.`group_name` = x5.`NAME`)
        LEFT OUTER JOIN
    `WORKSPACE_USER_SHARE` x4 ON x3.`SUBJECT_ID` = x4.`user_subject_id`;
```

explain plan:

screen shot 2017-01-25 at 9 31 03 am

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 13]
```diff
@@ -496,12 +573,23 @@ trait WorkspaceComponent {
...
+        (subjectId, accessLevel, permission.isDefined)
+      }
+      val groupsWithSharePermission = subGroupQuery.joinLeft(workspaceGroupShareQuery).on(_._1 === _.groupName).map { case ((groupName, accessLevel), permission) =>
+        (groupName, accessLevel, permission.isDefined)
+      }
+
       for {
-        users <- userQuery.result.map {
-          _.map { case (subjectId, accessLevel) => (Left(RawlsUserRef(RawlsUserSubjectId(subjectId))), WorkspaceAccessLevels.withName(accessLevel)) }
+        users <- usersWithSharePermission.result.map {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:580]: do this map and the one for groups a single time in the yield

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 14]
```diff
@@ -242,25 +267,73 @@ trait WorkspaceComponent {
...
 
-      (accessAndUserEmail union accessAndSubGroupEmail).result.map(_.map { case (access, email) => 
-        (email, WorkspaceAccessLevels.withName(access))
-      })
+      /*  The left join here is important. Since we are only going to store share-permission records for users and groups that have been
+          explicitly granted that ability, we want to be able to return all users and groups even if they don't have share permissions.
+          After the join, a null column equates to false, and a non-null column equates to true. This conversion is done with permission.isDefined
+       */
+      val userShareQuery = accessAndUserEmail.joinLeft(workspaceUserShareQuery).on(_._3 === _.userSubjectId).map { case ((accessLevel, userEmail, _), shareRecord) => (accessLevel, userEmail, shareRecord.isDefined) }
+      val groupShareQuery = accessAndSubGroupEmail.joinLeft(workspaceGroupShareQuery).on(_._3 === _.groupName).map { case ((accessLevel, groupEmail, _), shareRecord) => (accessLevel, groupEmail, shareRecord.isDefined) }
```
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala]: The SQL query that slick actually generates:

```
SELECT 
    x2.`access_level`, x3.`EMAIL`, x4.`workspace_id` IS NOT NULL
FROM
    `WORKSPACE_ACCESS` x2
        INNER JOIN
    (`GROUP` x5
    INNER JOIN (`GROUP_SUBGROUPS` x6
    INNER JOIN `GROUP` x3 ON x3.`NAME` = x6.`CHILD_NAME`) ON x5.`NAME` = x6.`PARENT_NAME`) ON ((x2.`workspace_id` = 0xE5D5CA221F0146DF9F92551963A9A1)
        AND (x2.`is_realm_acl` = FALSE))
        AND (x2.`group_name` = x5.`NAME`)
        LEFT OUTER JOIN
    `WORKSPACE_GROUP_SHARE` x4 ON x3.`NAME` = x4.`group_name`;
```

explain plan:

screen shot 2017-01-25 at 9 32 22 am

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 15]
```diff
@@ -507,15 +513,33 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
...
       val usersNotFound = emailsNotFound.filterNot(aclUpdate => invitesUpdated.map(_.email).contains(aclUpdate.email)).filterNot(aclUpdate => existingInvites.map(_.email).contains(aclUpdate.email))
+      val invitesSent = invitesUpdated.filterNot(aclUpdate => existingInvites.map(_.email).contains(aclUpdate.email))
 
-      WorkspaceACLUpdateResponseList(usersUpdated, invitesUpdated, usersNotFound)
+      WorkspaceACLUpdateResponseList(usersUpdated, invitesSent, invitesUpdated.diff(invitesSent), usersNotFound)
+    }
+
+    def updateWorkspaceSharePermissions(actualChangesToMake: Map[Either[RawlsUserRef, RawlsGroupRef], Option[Boolean]]) = {
+      val (usersToAdd, usersToRemove) = actualChangesToMake.collect{ case (Left(userRef), canShare) => userRef -> canShare }
+        .filter { case (_, canShare) => canShare.isDefined }.map{ case (user, canShare) => user -> canShare.get }.toSeq.partition { case (_, canShare) => canShare }
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: you can get rid of the inner filter and map
``` 
     val (usersToAdd, usersToRemove) = actualChangesToMake.collect{ case (Left(userRef), Some(canShare)) => userRef -> canShare }.toSeq.partition { case (_, canShare) => canShare }
```

**Review by dvoet — COMMENTED** [review]
last change

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 16]
```diff
@@ -473,7 +479,7 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
       dataSource.inTransaction { dataAccess =>
         withWorkspaceContext(workspaceName, dataAccess) { workspaceContext =>
           dataAccess.workspaceQuery.getInvites(workspaceContext.workspaceId).map { pairs =>
-            pairs.map(pair => WorkspaceACLUpdate(pair._1, pair._2))
+            pairs.map(pair => WorkspaceACLUpdate(pair._1, pair._2, None))
```
Comment (breilly2) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Can `WorkspaceACLUpdate` default the 3rd parameter to `None`? That would make this easier to read by not having to wonder what `None` here means.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala [hunk 17]
```diff
@@ -32,6 +32,9 @@ case class WorkspaceAccessRecord(workspaceId: UUID, groupName: String, accessLev
 /** result class for workspaceQuery.findAssociatedGroupsToIntersect, target = group1 intersect group2 */
 case class GroupsToIntersect(target: RawlsGroupRef, group1: RawlsGroupRef, group2: RawlsGroupRef)
 
+case class WorkspaceUserShareRecord(workspaceId: UUID, subjectId: String)
+case class WorkspaceGroupShareRecord(workspaceId: UUID, groupName: String)
```
Comment (breilly2) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:36]: It bothers me that there are separate tables here. I guess it makes sense that `RawlsUser` and `RawlsGroup` are different entities. However, there is definitely an area of functional overlap that is leading to a lot of code duplication. This duplication is now extending to the database for a feature that is truly shared between the two.

I don't have any immediate suggestions for changing this, and maybe there's something forcing it to be this way. However, this does seem to be the root of quite a bit of code duplication in these changes and I wanted to point it out so we could keep it in mind going forward.

**Review by breilly2 — CHANGES_REQUESTED** [review]
Also awaiting implementations for some tests, but wanted to get a batch of comments in.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 18]
```diff
@@ -1694,23 +1695,78 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
...
+        assertResult(StatusCodes.OK, response.entity.asString) { status }
+        responseAs[WorkspaceACL].acl should contain (testData.userOwner.userEmail.value -> AccessEntry(WorkspaceAccessLevels.ProjectOwner, false, true))
+      }
   }
 
-  it should "allow a user in a group with share permissions to share equal to and below their access level" in withTestDataApiServicesAndUser("writer-access") { services =>
-    assert(true)
+  it should "allow a user in a group with share permissions to share equal to and below their access level" in withTestDataApiServicesAndUser("reader-access") { services =>
+    import WorkspaceACLJsonSupport._
+    Patch(s"/workspaces/${testData.workspaceToTestGrant.namespace}/${testData.workspaceToTestGrant.name}/acl", httpJson(Seq(WorkspaceACLUpdate(testData.userReader2.userEmail.value, WorkspaceAccessLevels.Read, None)))) ~>
```
Comment (breilly2) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: I'm not 100% sure this is a different test from "allow a writer with share permissions to share equal to and below their access level". Isn't `testData.userReader2` the target of the change, not the user making the change? I would expect there to be a difference with `in withTestDataApiServiceAndUser("reader-access")` to have the change being requested by a user in a group.

**Review by breilly2 — CHANGES_REQUESTED** [review]

File: core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala (hunk 18 above)
Comment (MatthewBemis) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: This test is a little tricky because I had to set part of it up in TestDriverComponent. I wanted to test that a user in dbGapAuthorizedUsers (user "reader-access") could share if they were in that group and the group had share permission on the workspace.

Relevant lines are 209 of TestDriverComponent where I add "reader-access" to the dbGap group and line 544 where I manipulate the database to give that group the share permission on the test workspace.

The reason that I had to do it this way is that I can only run a test as one user. I'd need to use the user "owner-access" in order to add dbGapAuthorizedUsers to the workspace with the share permissions, but then I can't test anything as the user "reader-access". 

Basically, the assumption for this test is that dbGapAuthorizedUsers has already been granted share permissions by the owner and that "reader-access" is a member of that group.

Comment (breilly2) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: Okay, I understand that now.

The problem I see is that the group membership and permissions are not obvious when looking at this test. You have to trace through the all-in-one test fixture to see it, including understanding that there is a database lookup linking `"reader-access"` to `testData.userReader` because there is no way to navigate that statically. The risk with having these facts so far apart is that it's not obvious how changes in `TestDriverComponent` might affect which tests. For example, I committed tests related to bucket usage that use `"reader-access"` on the same day that you added that user to a group. So, while I was doing development, the user wasn't in a group but, by the time both of our branches made it to develop, the user was. Thankfully, it doesn't make a difference for my tests, but I think it illustrates the danger that I talked about at the cross-team tech meeting a few weeks ago.

Can the username be changed to make the user specifics obvious, such as `"reader-access-group-can-share"`? Or maybe even introduce a new user with that name, leaving `"reader-access"` as it was.

**Review by breilly2 — APPROVED** [review]
:+1: Awesome!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala (hunk 17 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala:36]: agreed

**Review by dvoet — APPROVED** [review]
